vscode/qodana/src/core/cloud/link.ts

```diff
@@ -0,0 +1,99 @@
+import * as vscode from "vscode";
+import {CONF_PROJ_ID, STATE_LINKED} from "../config";
+import {extensionInstance} from "../extension";
+import {CloudProjectResponse} from "./api";
+import {openReportByProjectId} from "../report";
+import {Events} from "../events";
+import telemetry from "../telemetry";
+
+export class LinkService {
+    private linkedProjectId: string | undefined;
+    private isLinked: boolean = false;
+    private projectProperties: CloudProjectResponse | undefined;
+
+    constructor(private context: vscode.ExtensionContext) {
+        vscode.workspace.onDidChangeConfiguration(async (e) => {
+            if (e.affectsConfiguration('qodana')) {
+                await this.selectAndLink();
+            }
+        });
+        this.selectAndLink().then();
+    }
+
+    private async selectAndLink() {
+        let projectId = vscode.workspace.getConfiguration().get<string>(CONF_PROJ_ID);
+        if (projectId) {
+            this.selectProject(projectId);
+            await this.linkProject();
+        } else {
+            await this.unlinkProject();
+        }
+    }
+
+    selectProject(projectId: string) {
+        this.linkedProjectId = projectId;
+    }
+
+    async linkProject() {
+        let projectId = this.linkedProjectId;
+        if (projectId === undefined) {
+            return;
+        }
+        let projectProperties = await this.getProjectProperties(projectId);
+        if (!projectProperties) {
+            return;
+        }
+        this.projectProperties = projectProperties;
+        this.isLinked = true;
+        vscode.workspace.getConfiguration().update(CONF_PROJ_ID, this.linkedProjectId, vscode.ConfigurationTarget.Workspace);
+        vscode.commands.executeCommand('setContext', STATE_LINKED, true);
+        telemetry.projectLinked();
+        Events.instance.fireProjectLinked();
+    }
+
+    async unlinkProject() {
+        this.isLinked = false;
+        this.projectProperties = undefined;
+        vscode.workspace.getConfiguration().update(CONF_PROJ_ID, undefined, vscode.ConfigurationTarget.Workspace);
+        vscode.commands.executeCommand('setContext', STATE_LINKED, false);
+        telemetry.projectUnlinked();
+        await extensionInstance.closeReport();
```

Comment:
Should we fire event in this case?

Reply:
We use only projectLinked event for UI for now. I can add separate event on unlink, but not sure if we need it